### DISCUSSION TITLE
Move jails to different dataset

### DISCUSSION THREAD

**Discussion by MildlyInterested** [Jip-Hop/jailmaker#105] [Q&A]
Hello,

I've been in the process of moving my jailmaker dataset to a different drive/pool and have been encountering an issue.
Jails created on the old location do not start anymore, logs say this:
```
Mar 24 15:58:41 truenas systemd-nspawn[18505]: execv(/usr/lib/systemd/systemd, /lib/systemd/systemd, /sbin/init) failed: No such file or directory
Mar 24 15:58:41 truenas systemd[1]: jlmkr-docker-jail.service: Main process exited, code=exited, status=1/FAILURE
```
Jails that have been created after the move are starting fine, so I'm guessing there's some old paths hardcoded _somehwere_ but I don't know where.

**Comment by wolffstarr** [comment]
I actually just went through this a few minutes ago - how did you move datasets? I did a replication (my goal was to get the dataset onto a different pool moreso than changing datasets) and updated my init script to point at the new dataset.

At a guess, did you move the jail directories for the old jails to the new dataset? The script is likely going to operate relative to the directory it's in, so if the older jails are in the old dataset, the script doesn't know about them.

**Reply by MildlyInterested** [reply]
I copied all contents of the old dataset with rsync to my new one, including the jail directories.
It does find the moved jails, it just can't start them because of the error I pasted above. Editing the jail configs results in different behaviour, so it's definitely reading the configs.
`jlmkr list` lists the moved jails correctly.

**Reply by wolffstarr** [reply]
If you run `jlmkr log <JailName>` does that have anything more detailed?

As far as I knew (and I'm by no means an expert) the only thing that puts a persistence from the script in place is `jlmkr startup`, but that should only be a shell alias and (if possible) a symlink - everything else should be based on the jailmaker dataset it's running from.

**Reply by MildlyInterested** [reply]
Thanks for helping out, the original snippet is the only error type output I got from `jlmkr log`, all the rest of the log looked exactly like logs of working jails. 
My server is sadly out of commission right now so I can't really check, hopefully end of the week I can test.

**Reply by Lockszmith-GH** [reply]
Just an FYI, datasets can be renamed - as long as they are not activley used. Make sure everyhting is stopped and cd out of the tree.

**Reply by Jip-Hop** [reply]
I think your rsync operation went wrong. Maybe you didn't preserve file permissions (executable bit etc.). The logs show that jailmaker is running systemd-nspawn which is then trying to start the init system inside the jail (systemd). But it can't find the `sbin/init` file inside the rootfs folder inside the jail directory. So the work jailmaker is doing is working properly as far as I can tell.

**Reply by MildlyInterested** [reply]
I didn't even think about permissions! It's most likely that, I'll report back once my server is back up, thanks!

**Reply by Neurrone** [reply]
Hey, did you ever figure this out? I'm thinking of doing the same thing, to migrate my jails from a zpool with HDDs to an SSD pool.

**Reply by Lockszmith-GH** [reply]
@Neurrone If you're transfering from one pool to another, you can use zfs send and recieve, that's probably the best way to preserve the pool including any snapshot it might have. (If you don't want the snapshots you can destroy them before the transfer)

**Comment by MildlyInterested** [comment]
Took *slightly* longer than anticipated but it was probably a permissions issue.
Ended up transferring the whole pool with zfs send|receive (https://www.truenas.com/community/threads/copy-move-dataset.28876/) and the jails started without issue!
